sdks/java/io/google-ads/src/main/java/org/apache/beam/sdk/io/googleads/GoogleAdsV14.java

```diff
@@ -0,0 +1,647 @@
+      checkArgumentNotNull(customerId, "customerId cannot be null");
+
+      return withCustomerIds(ImmutableList.of(customerId));
+    }
+
+    /**
+     * Creates and returns a new {@link Read} transform with the specified query. The query will be
+     * executed for each customer ID.
+     *
+     * @param query
+     * @return A new {@link Read} transform with the specified query.
+     * @see SearchGoogleAdsStreamRequest
+     * @see #withCustomerId(Long)
+     * @see #withCustomerIds(List)
+     */
+    public Read withQuery(String query) {
+      checkArgumentNotNull(query, "query cannot be null");
+      checkArgument(!query.isEmpty(), "query cannot be empty");
+
+      return toBuilder().setQuery(query).build();
+    }
+
+    /**
+     * Creates and returns a new {@link Read} transform with the specified client factory. A {@link
+     * GoogleAdsClientFactory} builds the {@link GoogleAdsClient} used to construct service clients.
+     * The {@link DefaultGoogleAdsClientFactory} should be sufficient for most purposes unless the
+     * construction of {@link GoogleAdsClient} requires customization.
+     *
+     * @param googleAdsClientFactory
+     * @return A new {@link Read} transform with the specified client factory.
+     * @see GoogleAdsClient
+     */
+    public Read withGoogleAdsClientFactory(GoogleAdsClientFactory googleAdsClientFactory) {
+      checkArgumentNotNull(googleAdsClientFactory, "googleAdsClientFactory cannot be null");
+
+      return toBuilder().setGoogleAdsClientFactory(googleAdsClientFactory).build();
+    }
+
+    /**
+     * Creates and returns a new {@link Read} transform with the specified rate limit policy
+     * factory. A {@link RateLimitPolicyFactory} builds the {@link RateLimitPolicy} used to limit
+     * the number of requests made by {@link ReadAll.ReadAllFn}. The Google Ads API enforces global
+     * limits from the developer token down to the customer ID and it is recommended to host a
+     * shared rate limiting service to coordinate traffic to the Google Ads API across all
+     * applications using the same developer token. Users of these transforms are strongly advised
+     * to implement their own {@link RateLimitPolicy} and {@link RateLimitPolicyFactory} to interact
+     * with a shared rate limiting service for any production workloads.
+     *
+     * @param rateLimitPolicyFactory
+     * @return A new {@link Read} transform with the specified rate limit policy factory.
+     * @see GoogleAdsClient
+     */
+    public Read withRateLimitPolicy(RateLimitPolicyFactory rateLimitPolicyFactory) {
+      checkArgumentNotNull(rateLimitPolicyFactory, "rateLimitPolicyFactory cannot be null");
+
+      return toBuilder().setRateLimitPolicyFactory(rateLimitPolicyFactory).build();
+    }
+
+    @Override
+    public PCollection<GoogleAdsRow> expand(PBegin input) {
+      String query = getQuery();
+      List<Long> customerIds = getCustomerIds();
```

Comment:
This pattern is limiting. Because the customer IDs are specified at pipeline construction time, they cannot be dynamic. 

Its likely worth reworking this IO to instead consume a PCollection of those IDs, so that they can be dynamically generated by some upstream source.

Reply:
I've changed the transform to take a `PCollection<String>` as input. `Read` can now be used to issue a single query for many customer IDs and `ReadAll` still provides the flexibility to alter the query per element (e.g. to support queries of the same output shape with differing WHERE clauses per customer ID).

Customer IDs are typically expressed as `long`, including in the Google Ads client library example code, but the protobuf schema for API requests defines customer IDs as `String` (oddly enough, linkedCustomerId/loginCustomerId are both `long` in the client...). To avoid baking assumptions about the identifier's format into the IO I think it makes sense to use `String` as the input type. The documentation example reflects this as well.